segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,407 @@
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return supplyAsync(targetHandle.getSegmentName(),
+                    () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+        this.closed.getAndSet(true);
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    @SneakyThrows(StreamSegmentNotExistsException.class)
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        }
+
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return FileSystemSegmentHandle.readHandle(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            LoggerHelpers.traceLeave(log, "openRead", traceId);
+            return FileSystemSegmentHandle.writeHandle(streamSegmentName);
+        } else {
+            LoggerHelpers.traceLeave(log, "openWrite", traceId);
+            return openRead(streamSegmentName).get();
```

Comment:
I left a similar comment somewhere else in this file. You are essentially blocking the current thread here while `openRead` is executing asynchronously. For a thread pool of 20, 10 of such calls done concurrently will use up the entire thread pool and lock up the entire application. 

Please rewrite the code to not do this.